wagtail/contrib/modeladmin/helpers/button.py

```diff
@@ -107,6 +107,17 @@ def get_buttons_for_obj(
             btns.append(self.delete_button(pk, classnames_add, classnames_exclude))
         return btns
 
+    def get_primary_button_for_obj(self, obj):
+        ph = self.permission_helper
+        usr = self.request.user
+        pk = getattr(obj, self.opts.pk.attname)
+        if ph.user_can_inspect_obj(usr, obj):
+            return self.inspect_button(pk)
+        elif ph.user_can_edit_obj(usr, obj):
+            return self.edit_button(pk)
+        else:
+            return None
```

Comment:
One thing that I think is important to change here is the order of these conditionals. If you had both edit AND inspect permission, I think user expectation would be that you were taken to the **edit** view over the **inspect** view. So, we should make the edit permission check first.

I've also included a couple of optional tweaks in the below suggestion:

Firstly, one of the nicest thing about methods that `return` a single value is that the code terminates at the point of return. This allows you to think about the subsequent lines differently. You know that, if the code is running, the preceding condition MUST have failed. You can use this knowledge to eliminate `if` / `elif` / `else`, making code easier to read (the brain cannot sufficiently reason about an `elif` without also understanding the failed `if`/`elif` statements that came before).

Secondly, in Python, if a function/method hasn't returned anything by the time the last line if executed, `None` is returned automatically; So adding either (functionally equivalent) lines to the end of the function isn't necessary:

```
return
```

```
return None
```

Explicitly returning `None` isn't necessarily a bad thing though, as it arguably makes code easier to read for developers who haven't quite committed that knowledge to memory.

```suggestion
        if ph.user_can_edit_obj(usr, obj):
            return self.edit_button(pk)
        if ph.user_can_inspect_obj(usr, obj):
            return self.inspect_button(pk)
```

Reply:
> One thing that I think is important to change here is the order of these conditionals. If you had both edit AND inspect permission, I think user expectation would be that you were taken to the **edit** view over the **inspect** view. So, we should make the edit permission check first.

My thought was, that the inspect view has to be enabled by the developer explicitly, so even with edit permissions, it could be prioritized. But looking at it again, changing the order also makes sense.

> Firstly, one of the nicest thing about methods that `return` a single value is that the code terminates at the point of return. This allows you to think about the subsequent lines differently. You know that, if the code is running, the preceding condition MUST have failed. You can use this knowledge to eliminate `if` / `elif` / `else`, making code easier to read (the brain cannot sufficiently reason about an `elif` without also understanding the failed `if`/`elif` statements that came before).

You're right, in that specific case it's actually smarter that way, so future sub-branches inside the `if`, which won't return, actually allow a fallback to another button.
I personally prefer the `if` / `elif` / `else` in first approaches as its more restricted and for me more obvious, but probably I'm a bit paranoid due too many security-related code reviews.

> Explicitly returning `None` isn't necessarily a bad thing though, as it arguably makes code easier to read for developers who haven't quite committed that knowledge to memory.

I prefer explicit return types, as it makes the developer's intention more clear and so it's easier to differentiate from forgotten return values in code reviews, I think.